relayer/processor/path_processor_internal.go

```diff
@@ -831,137 +852,380 @@ func (pp *PathProcessor) assembleAndSendMessages(
-		// only assemble and send channel handshake messages if there are no conn handshake messages
-		// this prioritizes connection handshake messages, useful if a connection handshake needs to occur before a channel handshake
-		for i, msg := range messages.channelMessages {
-			wg.Add(1)
-			go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
+	assembled := 0
+	for _, m := range om.connMsgs {
+		if m.m != nil {
+			assembled++
+		}
+	}
+	for _, m := range om.chanMsgs {
+		if m.m != nil {
+			assembled++
 		}
+	}
+	for _, m := range om.clientICQMsgs {
+		if m.m != nil {
+			assembled++
+		}
+	}
+	for _, m := range om.pktMsgs {
+		if m.m != nil {
+			assembled++
+		}
+	}
+
+	broadcastBatch := dst.chainProvider.ProviderConfig().BroadcastMode() == provider.BroadcastModeBatch
+	batchMsgs := []provider.RelayerMessage{om.msgUpdateClient}
+
+	for _, t := range om.connMsgs {
+		retries := dst.trackProcessingConnectionMessage(t)
+		if t.m == nil {
+			continue
+		}
+		if broadcastBatch && retries == 0 {
+			batchMsgs = append(batchMsgs, t.m)
+			continue
+		}
+		go pp.sendConnectionMessage(ctx, src, dst, om.msgUpdateClient, t)
+	}
 
-		wg.Wait()
+	for _, t := range om.chanMsgs {
+		retries := dst.trackProcessingChannelMessage(t)
+		if t.m == nil {
+			continue
+		}
+		if broadcastBatch && retries == 0 {
+			batchMsgs = append(batchMsgs, t.m)
+			continue
+		}
+		go pp.sendChannelMessage(ctx, src, dst, om.msgUpdateClient, t)
 	}
 
-	if len(om.msgs) == 1 {
-		om.clientICQMsgs = make([]clientICQMessageToTrack, len(messages.clientICQMessages))
-		// only assemble and send ICQ messages if there are no conn or chan handshake messages
-		for i, msg := range messages.clientICQMessages {
-			wg.Add(1)
-			go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
+	for _, t := range om.clientICQMsgs {
+		retries := dst.trackProcessingClientICQMessage(t)
+		if t.m == nil {
+			continue
+		}
+		if broadcastBatch && retries == 0 {
+			batchMsgs = append(batchMsgs, t.m)
+			continue
 		}
+		go pp.sendClientICQMessage(ctx, src, dst, om.msgUpdateClient, t)
 
-		wg.Wait()
 	}
 
-	if len(om.msgs) == 1 {
-		om.pktMsgs = make([]packetMessageToTrack, len(messages.packetMessages))
-		// only assemble and send packet messages if there are no handshake messages
-		for i, msg := range messages.packetMessages {
-			wg.Add(1)
-			go pp.assembleMessage(ctx, msg, src, dst, &om, i, &wg)
+	for _, t := range om.pktMsgs {
+		retries := dst.trackProcessingPacketMessage(t)
+		if t.m == nil {
+			continue
 		}
+		if broadcastBatch && retries == 0 {
+			batchMsgs = append(batchMsgs, t.m)
+			continue
+		}
+		go pp.sendPacketMessage(ctx, src, dst, om.msgUpdateClient, t)
+	}
 
-		wg.Wait()
+	if len(batchMsgs) > 1 {
+		go pp.sendBatchMessages(ctx, src, dst, batchMsgs, om.pktMsgs)
 	}
 
-	if len(om.msgs) == 1 && !needsClientUpdate {
+	if assembled == 0 {
+		if needsClientUpdate {
+			go pp.sendClientUpdate(ctx, src, dst, om.msgUpdateClient)
+			return nil
+		}
```

Comment:
nit: if we inverse the check on L969 we can remove the nesting here

Reply:
good call, handled in `messageProcessor` in `trackAndSendMessages`